components/file_system/src/rate_limiter.rs

```diff
@@ -505,39 +353,82 @@ impl IORateLimiter {
+    /// treated as physical I/Os without further scrutiny.
+    pub async fn async_request(&self, io_type: IOType, bytes: usize) -> usize {
+        let priority = IOPriority::unsafe_from_u32(
+            self.priority_map[io_type as usize].load(Ordering::Relaxed),
+        );
+        request_physical_imp!(self, priority, bytes, async);
         bytes
     }
 
     #[cfg(test)]
-    fn request_with_skewed_clock(&self, io_type: IOType, io_op: IOOp, mut bytes: usize) -> usize {
+    fn request_for_test(&self, io_type: IOType, io_op: IOOp, bytes: usize) -> usize {
         if self.mode.contains(io_op) {
-            bytes = self.throughput_limiter.request_with_skewed_clock(
-                IOPriority::unsafe_from_u32(
-                    self.priority_map[io_type as usize].load(Ordering::Relaxed),
-                ),
-                bytes,
+            let priority = IOPriority::unsafe_from_u32(
+                self.priority_map[io_type as usize].load(Ordering::Relaxed),
             );
+            request_physical_imp!(self, priority, bytes, sync);
         }
         if let Some(stats) = &self.stats {
             stats.record(io_type, io_op, bytes);
         }
         bytes
     }
+
+    /// Updates and refills I/O budgets for next epoch based on I/O priority.
+    /// This method is a no-op when high-priority I/O rate limit equals zero.
+    /// Here we provide best-effort priority control:
+    /// 1) Limited I/O budget is assigned to lower priority to ensure higher priority can at least
+    ///    consume the same I/O amount as the last few epochs without breaching global threshold.
+    /// 2) Higher priority may temporarily use lower priority's I/O budgets. When this happens,
+    ///    total I/O flow could exceed global threshold.
+    /// 3) Highest priority I/O alone must not exceed global threshold (in strict mode).
+    fn refill(&self, locked: &mut IORateLimiterInner, now: Instant) {
+        let mut budgets = self.bytes_per_epoch.load(Ordering::Relaxed);
+        if budgets == 0 {
+            // Rate limit is toggled off in the meantime.
+            return;
+        }
+        let new_refill_time = now + DEFAULT_REFILL_PERIOD;
+        let elapsed_epochs = (new_refill_time - locked.next_refill_time).as_secs_f32()
+            / DEFAULT_REFILL_PERIOD.as_secs_f32();
+        locked.next_refill_time = new_refill_time;
+
+        for pri in &[IOPriority::High, IOPriority::Medium, IOPriority::Low] {
+            let p = *pri as usize;
+
+            match *pri {
+                IOPriority::Medium => RATE_LIMITER_MAX_BYTES_PER_SEC
+                    .medium
+                    .set((budgets * DEFAULT_REFILLS_PER_SEC) as i64),
+                IOPriority::Low => {
+                    RATE_LIMITER_MAX_BYTES_PER_SEC
+                        .low
+                        .set((budgets * DEFAULT_REFILLS_PER_SEC) as i64);
+                }
+                _ => {}
```

Comment:
seem that `IOPriority::High` always be bytes_per_sec here?

Reply:
Yes.